<a href="https://colab.research.google.com/github/TithiSreemany/Artificial-Neural-Network-ANN-/blob/main/Artificial_Neural_Network.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Artificial Neural Network (ANN)


### Importing the libraries

In [2]:
import numpy as np
import pandas as pd
import tensorflow as tf


In [3]:
tf.__version__

'2.3.0'

## Part 1 - Data Preprocessing

## Importing the dataset

In [4]:
dataset = pd.read_csv('Churn_Modelling.csv')
X = dataset.iloc[:, 3:-1].values
y = dataset.iloc[:, -1].values

In [5]:
print(X)

[[619 'France' 'Female' ... 1 1 101348.88]
 [608 'Spain' 'Female' ... 0 1 112542.58]
 [502 'France' 'Female' ... 1 0 113931.57]
 ...
 [709 'France' 'Female' ... 0 1 42085.58]
 [772 'Germany' 'Male' ... 1 0 92888.52]
 [792 'France' 'Female' ... 1 0 38190.78]]


In [6]:
print(y)

[1 0 1 ... 1 1 0]


### Encoding categorical data

Label Encoding the "Gender" column

In [7]:
from sklearn.preprocessing import LabelEncoder
le= LabelEncoder()
X[:,2]=le.fit_transform(X[:,2])

In [8]:
print(X)

[[619 'France' 0 ... 1 1 101348.88]
 [608 'Spain' 0 ... 0 1 112542.58]
 [502 'France' 0 ... 1 0 113931.57]
 ...
 [709 'France' 0 ... 0 1 42085.58]
 [772 'Germany' 1 ... 1 0 92888.52]
 [792 'France' 0 ... 1 0 38190.78]]


One Hot Encoding the "Geography" column

In [9]:
from sklearn.compose import ColumnTransformer
from sklearn.preprocessing import OneHotEncoder
ct = ColumnTransformer(transformers=[('encoder', OneHotEncoder(), [1])], remainder='passthrough' ) 
X = np.array(ct.fit_transform(X))

In [10]:
print(X)

[[1.0 0.0 0.0 ... 1 1 101348.88]
 [0.0 0.0 1.0 ... 0 1 112542.58]
 [1.0 0.0 0.0 ... 1 0 113931.57]
 ...
 [1.0 0.0 0.0 ... 0 1 42085.58]
 [0.0 1.0 0.0 ... 1 0 92888.52]
 [1.0 0.0 0.0 ... 1 0 38190.78]]


### Splitting the dataset into the Training set and Test set

In [12]:
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 0)

### Feature Scaling

In [13]:
from sklearn.preprocessing import StandardScaler
sc = StandardScaler()
X_train = sc.fit_transform(X_train)
X_test = sc.transform(X_test)

## Part 2 - Building the ANN

### Initialising the ANN

In [20]:
ann= tf.keras.models.Sequential()

### Adding the input layer and the first hidden layer

In [21]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the second hidden layer

In [22]:
ann.add(tf.keras.layers.Dense(units=6, activation='relu'))

### Adding the output layer

In [23]:
ann.add(tf.keras.layers.Dense(units=1, activation='sigmoid'))

## Part 3 - Training the ANN

### Compiling the ANN

In [24]:
ann.compile(optimizer='adam', loss='binary_crossentropy', metrics=['accuracy'])

### Training the ANN on the Training set

In [25]:
ann.fit(X_train, y_train, batch_size = 32, epochs = 100)

Epoch 1/100
250/250 [==============================] - 0s 928us/step - loss: 0.6263 - accuracy: 0.6398
Epoch 2/100
250/250 [==============================] - 0s 985us/step - loss: 0.4883 - accuracy: 0.7855
Epoch 3/100
250/250 [==============================] - 0s 937us/step - loss: 0.4560 - accuracy: 0.7940
Epoch 4/100
250/250 [==============================] - 0s 921us/step - loss: 0.4357 - accuracy: 0.8023
Epoch 5/100
250/250 [==============================] - 0s 889us/step - loss: 0.4158 - accuracy: 0.8173
Epoch 6/100
250/250 [==============================] - 0s 886us/step - loss: 0.3973 - accuracy: 0.8269
Epoch 7/100
250/250 [==============================] - 0s 1ms/step - loss: 0.3849 - accuracy: 0.8361
Epoch 8/100
250/250 [==============================] - 0s 857us/step - loss: 0.3769 - accuracy: 0.8401
Epoch 9/100
250/250 [==============================] - 0s 909us/step - loss: 0.3719 - accuracy: 0.8420
Epoch 10/100
250/250 [==============================] - 0s 934us/step - los

## Part 4 - Making the predictions and evaluating the model


### Predicting the result of a single observation

Use our ANN model to predict if the customer with the following informations will leave the bank:

Geography: France

Credit Score: 600

Gender: Male

Age: 40 years old

Tenure: 3 years

Balance: $ 60000

Number of Products: 2

Does this customer have a credit card? Yes

Is this customer an Active Member: Yes

Estimated Salary: $ 50000

So, should we say goodbye to that customer?

In [26]:
print(ann.predict(sc.transform([[1, 0, 0, 600, 1, 40, 3, 60000, 2, 1, 1, 50000]])) > 0.5)

[[False]]


Therefore, our ANN model predicts that this customer stays in the bank!

### Predicting the Test set results

In [27]:
y_pred = ann.predict(X_test)
y_pred = (y_pred > 0.5)
print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

[[0 0]
 [0 1]
 [0 0]
 ...
 [0 0]
 [0 0]
 [0 0]]


### Making the Confusion Matrix

In [28]:
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)

[[1499   96]
 [ 180  225]]


0.862

Hence we can conclude that our ANN Model has an accuracy of 86.2% in predicting the outcomes.